### ISSUE TITLE
`Mapping[Hashable, ...]` and `Mapping[Any, ...]` behaves differently

### ISSUE THREAD

**Issue by cmp0xff** [astral-sh/ty#3956]
### Summary

In pandas-dev/pandas-stubs#1809, I found the following issue:

```py
from collections.abc import Callable, Hashable, Mapping
from typing import Concatenate, Any

class NDArray: ...

def mean(_: NDArray, *args: Any, **kwargs: Any) -> Any:
    pass

def f(_: Mapping[Hashable, list[Callable[Concatenate[NDArray, ...], Any]]]) -> None:
    pass

def g(_: Mapping[Any, list[Callable[Concatenate[NDArray, ...], Any]]]) -> None:
    pass

f({"a": [mean]})  # false negative by ty
g({"a": [mean]})  # passes
```

- https://play.ty.dev/81d378a9-6d7d-40f5-bdca-98f4adb50be5
- https://gist.github.com/mypy-play/a537202d8b6dfcd3a2bd6ab6d4733bb5

### Version

20740c5d7

**Comment by jelle-openai** [comment]
Simplified repro:

```python
from collections.abc import Callable, Hashable, Mapping
from typing import Any

def mean(*args: Any, **kwargs: Any) -> Any:
    pass

def f(_: Mapping[Hashable, list[Callable[..., Any]]]) -> None:
    pass

def g(_: Mapping[Any, list[Callable[..., Any]]]) -> None:
    pass

f({"a": [mean]})  # error
g({"a": [mean]})  # passes
```

Produces ```Expected `Mapping[Hashable, list[(...) -> Any]]`, found `dict[Hashable, list[(...) -> Any]]` ```. No longer reproes if you use a list of strs, or if you use Callable directly.
